docker中的agent启动报index out of range错误

已经取得问题分析：agent运行在docker容器里，docker容器限制了cpu个数，agent拿到了正确的cpu个数，比如宿主机是32个核，限制docker容器为2个核，agent就拿到了cpu核数为2 。于是准备了一个长度为2的数组来放置各个cpu的性能数据。但是，接下来agent要去读取/proc/stat，docker容器内部看到的/proc/stat文件内容和宿主机一般无二，于是看到了32个核，于是index out of range。
那么，有什么解决的办法吗？

2015/11/10 16:36:43 cfg.go:120: read config file: cfg.json successfully
2015/11/10 16:36:43 http.go:74: listening :1988
panic: runtime error: index out of range

goroutine 17 [running]:
github.com/toolkits/nux.parseLine(0xc2080eb0c2, 0x3b, 0xf3e, 0xc208031680)
        /home/laiwei/gopath-open-falcon/src/github.com/toolkits/nux/cpustat.go:102 +0x2df
github.com/toolkits/nux.CurrentProcStat(0x0, 0x0, 0x0)
        /home/laiwei/gopath-open-falcon/src/github.com/toolkits/nux/cpustat.go:78 +0x38d
github.com/open-falcon/agent/funcs.UpdateCpuStat(0x0, 0x0)
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/funcs/cpustat.go:19 +0x31
created by main.main
goroutine 1 [select (no cases)]:
main.main()
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/main.go:49 +0x270

goroutine 7 [syscall, locked to thread]:
net._C2func_getaddrinfo(0x7fc6f80008c0, 0x0, 0xc208018dc8, 0xc208018d18, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/:26 +0x55
net.cgoLookupIPCNAME(0xc2080961c0, 0x15, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/cgo_unix.go:96 +0x1c5
net.lookupIP(0xc2080961c0, 0x15, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/lookup_unix.go:64 +0x5f
net.func路026(0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/lookup.go:79 +0x55
net.(*singleflight).doCall(0x9fc880, 0xc2080401e0, 0xc2080961c0, 0x15, 0xc20802ad90)
        /usr/local/go/src/net/singleflight.go:91 +0x2f
created by net.(*singleflight).DoChan
        /usr/local/go/src/net/singleflight.go:84 +0x42b

goroutine 18 [select]:
net.lookupIPDeadline(0xc2080961c0, 0x15, 0xecdd3a3c3, 0x2bdf936b, 0x9fd380, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/lookup.go:82 +0x6cb
        /usr/local/go/src/net/ipsock.go:285 +0x49b
        /usr/local/go/src/net/dial.go:110 +0x378
net.(*Dialer).Dial(0xc208031600, 0x809390, 0x3, 0xc2080961c0, 0x1a, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/dial.go:158 +0xf6
net.DialTimeout(0x809390, 0x3, 0xc2080961c0, 0x1a, 0x9502f9000, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/dial.go:150 +0xe6
github.com/toolkits/net.JsonRpcClient(0x809390, 0x3, 0xc2080961c0, 0x1a, 0x9502f9000, 0x0, 0x0, 0x0)
        /home/laiwei/gopath-open-falcon/src/github.com/toolkits/net/rpc.go:19 +0x71
github.com/open-falcon/agent/g.(*SingleConnRpcClient).insureConn(0xc208076c30)
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/g/rpc.go:39 +0xa1
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/g/rpc.go:61 +0x13b
github.com/open-falcon/agent/cron.reportAgentStatus(0xdf8475800)
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/reporter.go:32 +0x2d4
created by github.com/open-falcon/agent/cron.ReportAgentStatus
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/reporter.go:13 +0x72

goroutine 19 [sleep]:
github.com/open-falcon/agent/cron.syncBuiltinMetrics()
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/builtin.go:27 +0x77
created by github.com/open-falcon/agent/cron.SyncBuiltinMetrics
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/builtin.go:14 +0x56

goroutine 20 [sleep]:
github.com/open-falcon/agent/cron.syncTrustableIps()
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/ips.go:22 +0x4d
created by github.com/open-falcon/agent/cron.SyncTrustableIps
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/ips.go:12 +0x56

goroutine 21 [sleep]:
github.com/open-falcon/agent/cron.collect(0x3c, 0xc208086120, 0xb, 0xb)
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/collector.go:37 +0x3d
created by github.com/open-falcon/agent/cron.Collect
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/collector.go:29 +0xf7

goroutine 22 [sleep]:
github.com/open-falcon/agent/cron.collect(0x3c, 0xc2080b2030, 0x1, 0x1)
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/collector.go:37 +0x3d
created by github.com/open-falcon/agent/cron.Collect
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/collector.go:29 +0xf7

goroutine 23 [sleep]:
github.com/open-falcon/agent/cron.collect(0x3c, 0xc208070660, 0x2, 0x2)
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/collector.go:37 +0x3d
created by github.com/open-falcon/agent/cron.Collect
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/collector.go:29 +0xf7

goroutine 24 [sleep]:
github.com/open-falcon/agent/cron.collect(0x3c, 0xc2080b2038, 0x1, 0x1)
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/collector.go:37 +0x3d
created by github.com/open-falcon/agent/cron.Collect
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/cron/collector.go:29 +0xf7

goroutine 25 [IO wait]:
net.(*pollDesc).Wait(0xc2080105a0, 0x72, 0x0, 0x0)
        /usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc2080105a0, 0x0, 0x0)
        /usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc208010540, 0x0, 0x7fc70880bd58, 0xc2080706b8)
        /usr/local/go/src/net/fd_unix.go:419 +0x40b
net.(*TCPListener).AcceptTCP(0xc2080b2048, 0x529314, 0x0, 0x0)
        /usr/local/go/src/net/tcpsock_posix.go:234 +0x4e
net/http.tcpKeepAliveListener.Accept(0xc2080b2048, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/http/server.go:1976 +0x4c
net/http.(*Server).Serve(0xc208086180, 0x7fc70880d568, 0xc2080b2048, 0x0, 0x0)
        /usr/local/go/src/net/http/server.go:1728 +0x92
net/http.(*Server).ListenAndServe(0xc208086180, 0x0, 0x0)
        /usr/local/go/src/net/http/server.go:1718 +0x154
github.com/open-falcon/agent/http.Start()
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/http/http.go:75 +0x233
created by main.main
        /home/laiwei/gopath-open-falcon/src/github.com/open-falcon/agent/main.go:47 +0x269

goroutine 33 [syscall, locked to thread]:
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

--- a/cpustat.go
+++ b/cpustat.go
@@ -95,7 +95,7 @@ func parseLine(line []byte, ps *ProcStat) {
 
 	if strings.HasPrefix(fieldName, "cpu") {
 		idx, err := strconv.Atoi(fieldName[3:])
-		if err != nil {
+		if err != nil || idx >= len(ps.Cpus) {
 			return
 		}
 

